components/support/sql/src/conn_ext.rs

```diff
@@ -213,3 +217,72 @@ impl<'conn> ConnExt for UncheckedTransaction<'conn> {
         &*self
     }
 }
+
+/// This transaction is suitable for when a transaction is being used only
+/// for performance when updating a larger number of rows (ie, when the
+/// transaction is not necessary for data integrity). You can specify a
+/// duration for the maximum amount of time the transaction should be held.
+/// You should regularly call .maybe_commit() as part of your processing, and
+/// if the current transaction has been open for greater than the specified
+/// duration the transaction will be committed and another one started.
+/// You should always call .commit() at the end. Note that if you call
+/// .rollback() at the end, only the current transaction will be rolled back.
+pub struct TransactionForPerf<'conn> {
+    tx: UncheckedTransaction<'conn>,
+    behavior: TransactionBehavior,
+    commit_after: Duration,
+}
+
+impl<'conn> TransactionForPerf<'conn> {
+    /// Begin a new transaction which may be split into multiple transactions
+    /// for performance reasons. Cannot be nested, but this is not
+    /// enforced; use a rusqlite `savepoint` for nested transactions.
+    pub fn new(
+        conn: &'conn Connection,
+        behavior: TransactionBehavior,
```

Comment:
I wonder if we should consider enforcing this dynamically (e.g. by checking if a transaction is open somehow). That would avoid a TransactionForPerf committing a correctness transaction accidentally...

Reply:
You can do this using https://docs.rs/rusqlite/*/rusqlite/struct.Connection.html#method.is_autocommit. We should also probably do this in UncheckedTransaction too...

Reply:
Hmm, how could `TransactionForPerf` commit a correctness transaction accidentally? SQLite returns an error if you try to start a nested transaction, so, IIUC, `UncheckedTransaction::new` will fail?

Reply:
Ah, you're right.